File: codegen/typescript.go
package codegen

import (
	"github.com/openland/spacex-cli/il"
	"io/ioutil"
	"os"
	"path/filepath"
)

func GenerateTypescript(model *il.Model, to string) {
	output := NewOutput()

	setComma("'")
	setArgFunc("args")
	setNeedSemicolon(true)
	setExplicitArgs(true)

	output.WriteLine("/* tslint:disable */")
	output.WriteLine("/* eslint-disable */")
	output.WriteLine("// @ts-ignore")
	output.WriteLine("import { WebDefinitions, OperationDefinition, Definitions as AllDefinitions } from '@openland/spacex';")
	output.WriteLine("// @ts-ignore")
	output.WriteLine("const { list, notNull, scalar, field, obj, inline, fragment, args, fieldValue, refValue, intValue, floatValue, stringValue, boolValue, listValue, objectValue } = WebDefinitions;")
	output.WriteLine("")

	for _, f := range model.Fragments {
		output.NextScope()
		output.WriteLine("const " + f.Name + "Selector = ")
		output.IndentAdd()
		generateSelector(f.SelectionSet, output)
		output.IndentRemove()
		output.WriteLine("")
	}

	for _, f := range model.Queries {
		output.NextScope()
		output.WriteLine("const " + f.Name + "Selector = ")
		output.IndentAdd()
		generateSelector(f.SelectionSet, output)
		output.IndentRemove()
	}

	for _, f := range model.Mutations {
		output.NextScope()
		output.WriteLine("const " + f.Name + "Selector = ")
		output.IndentAdd()
		generateSelector(f.SelectionSet, output)
		output.IndentRemove()
	}

	for _, f := range model.Subscriptions {
		output.NextScope()
		output.WriteLine("const " + f.Name + "Selector = ")
		output.IndentAdd()
		generateSelector(f.SelectionSet, output)
		output.IndentRemove()
	}

	output.WriteLine("export const Operations: { [key: string]: OperationDefinition } = {")
	output.IndentAdd()
	for _, f := range model.Queries {
		output.WriteLine("" + f.Name + ": {")
		output.IndentAdd()
		output.WriteLine("kind: 'query',")
		output.WriteLine("name: '" + f.Name + "',")
		output.WriteLine("body: '" + f.Body + "',")
		output.WriteLine("selector: " + f.Name + "Selector")
		output.IndentRemove()
		output.WriteLine("},")
	}
	for _, f := range model.Mutations {
		output.WriteLine("" + f.Name + ": {")
		output.IndentAdd()
		output.WriteLine("kind: 'mutation',")
		output.WriteLine("name: '" + f.Name + "',")
		output.WriteLine("body: '" + f.Body + "',")
		output.WriteLine("selector: " + f.Name + "Selector")
		output.IndentRemove()
		output.WriteLine("},")
	}
	for _, f := range model.Subscriptions {
		output.WriteLine("" + f.Name + ": {")
		output.IndentAdd()
		output.WriteLine("kind: 'subscription',")
		output.WriteLine("name: '" + f.Name + "',")
		output.WriteLine("body: '" + f.Body + "',")
		output.WriteLine("selector: " + f.Name + "Selector")
		output.IndentRemove()
		output.WriteLine("},")
	}
	output.IndentRemove()
	output.WriteLine("};")

	output.WriteLine("export const Definitions: AllDefinitions = { operations: Operations };")

	// Write result
	err := os.MkdirAll(filepath.Dir(to), os.ModePerm)
	if err != nil {
		panic(err)
	}
	err = ioutil.WriteFile(to, []byte(output.String()), 0644)
	if err != nil {
		panic(err)
	}
}

File: test_data/output/typescript.ts
// @ts-ignore
import { list, notNull, scalar, field, obj, inline, fragment, args, fieldValue, refValue, intValue, floatValue, stringValue, boolValue, listValue, objectValue, OperationDefinition } from 'openland-graphql/spacex/types';

const DeviceNanoSelector = obj(
            field('__typename', '__typename', args(), notNull(scalar('String'))),
            field('id', 'id', args(), notNull(scalar('ID')))
        );

const LampShortSelector = obj(
            field('__typename', '__typename', args(), notNull(scalar('String'))),
            field('id', 'id', args(), notNull(scalar('ID'))),
            field('brightness', 'brightness', args(), notNull(scalar('Float'))),
            field('minBrightness', 'minBrightness', args(), notNull(scalar('Float'))),
            field('maxBrightness', 'maxBrightness', args(), notNull(scalar('Float')))
        );

const DeviceShortSelector = obj(
            field('__typename', '__typename', args(), notNull(scalar('String'))),
            field('description', 'description', args(), scalar('String')),
            field('addedBy', 'addedBy', args(), notNull(obj(
                    field('__typename', '__typename', args(), notNull(scalar('String'))),
                    field('id', 'id', args(), notNull(scalar('ID'))),
                    field('username', 'username', args(), notNull(scalar('String')))
                ))),
            fragment('Device', DeviceNanoSelector),
            inline('Lamp', obj(
                fragment('Lamp', LampShortSelector)
            )),
            inline('Switch', obj(
                field('id', 'id', args(), notNull(scalar('ID'))),
                field('numberOfButtons', 'numberOfButtons', args(), scalar('Int'))
            )),
            inline('Lamp', obj(
                field('id', 'id', args(), notNull(scalar('ID'))),
                field('brightness', 'brightness', args(), notNull(scalar('Float')))
            ))
        );

const UserShortSelector = obj(
            field('__typename', '__typename', args(), notNull(scalar('String'))),
            field('id', 'id', args(), notNull(scalar('ID')))
        );

const DiscoverDevicesSelector = obj(
            field('discover', 'discover', args(), notNull(list(notNull(obj(
                    field('id', 'id', args(), notNull(scalar('ID'))),
                    field('name', 'name', args(), notNull(scalar('String'))),
                    field('host', 'host', args(), notNull(scalar('String'))),
                    field('port', 'port', args(), notNull(scalar('Int')))
                )))))
        );
const AddUserSelector = obj(
            field('addUser', 'addUser', args(fieldValue("username", refValue('username')), fieldValue("password", refValue('password'))), notNull(obj(
                    fragment('User', UserShortSelector)
                )))
        );
const UpdateDeviceDescriptionSelector = obj(
            field('updateDeviceDescription', 'updateDeviceDescription', args(fieldValue("id", refValue('id')), fieldValue("description", refValue('description'))), notNull(obj(
                    fragment('Device', DeviceShortSelector)
                )))
        );
export const Operations: { [key: string]: OperationDefinition } = {
    DiscoverDevices: {
        kind: 'query',
        name: 'DiscoverDevices',
        body: 'query DiscoverDevices{discover{id name host port}}',
        selector: DiscoverDevicesSelector
    },
    AddUser: {
        kind: 'mutation',
        name: 'AddUser',
        body: 'mutation AddUser($username:String!,$password:String!){addUser(username:$username,password:$password){...UserShort}}fragment UserShort on User{__typename id}',
        selector: AddUserSelector
    },
    UpdateDeviceDescription: {
        kind: 'mutation',
        name: 'UpdateDeviceDescription',
        body: 'mutation UpdateDeviceDescription($id:ID!,$description:DeviceDescription!){updateDeviceDescription(id:$id,description:$description){...DeviceShort}}fragment DeviceShort on Device{__typename description addedBy{__typename id username}...DeviceNano ... on Lamp{...LampShort}... on Switch{id numberOfButtons}... on Lamp{id brightness}}fragment DeviceNano on Device{__typename id}fragment LampShort on Lamp{__typename id brightness minBrightness maxBrightness}',
        selector: UpdateDeviceDescriptionSelector
    },
};